File: src/test/java/uk/gov/justice/digital/hmpps/riskprofiler/controllers/ResourceTest.java
package uk.gov.justice.digital.hmpps.riskprofiler.controllers;

import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.json.JsonContent;
import org.springframework.boot.test.web.client.TestRestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.test.annotation.DirtiesContext;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;
import uk.gov.justice.digital.hmpps.riskprofiler.utils.JwtAuthenticationHelper;

import java.time.Duration;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
import static org.springframework.core.ResolvableType.forType;
import static org.springframework.http.HttpHeaders.ACCEPT;
import static org.springframework.http.HttpHeaders.AUTHORIZATION;
import static org.springframework.http.MediaType.APPLICATION_JSON_VALUE;

@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_CLASS)
@RunWith(SpringRunner.class)
@ActiveProfiles(profiles = {"test", "localstack"})
@SpringBootTest(webEnvironment = RANDOM_PORT)
public abstract class ResourceTest {

  @Autowired
  protected TestRestTemplate testRestTemplate;

  @Autowired
  protected JwtAuthenticationHelper jwtAuthHelper;

  HttpEntity<?> createHttpEntityWithBearerAuthorisation(final String user, final List<String> roles) {
    final var jwt = createJwt(user, roles);
    return createHttpEntity(jwt, null);
  }

  HttpEntity<?> createHttpEntity(final String bearerToken, final Object body) {
    final var headers = new HttpHeaders();
    headers.add(AUTHORIZATION, "Bearer " + bearerToken);
    headers.add(ACCEPT, APPLICATION_JSON_VALUE);
    if (body != null) {
      headers.add(HttpHeaders.CONTENT_TYPE, APPLICATION_JSON_VALUE);
    }
    return new HttpEntity<>(body, headers);
  }

  void assertThatStatus(final ResponseEntity<String> response, final int status) {
    assertThat(response.getStatusCodeValue()).withFailMessage("Expecting status code value <%s> to be equal to <%s> but it was not.\nBody was\n%s", response.getStatusCodeValue(), status, response.getBody()).isEqualTo(status);
  }

  void assertThatJsonFileAndStatus(final ResponseEntity<String> response, final int status, final String jsonFile) {
    assertThatStatus(response, status);
    assertThat(getBodyAsJsonContent(response)).isEqualToJson(jsonFile);
  }

  private <T> JsonContent<T> getBodyAsJsonContent(final ResponseEntity<String> response) {
    return new JsonContent<>(getClass(), forType(String.class), Objects.requireNonNull(response.getBody()));
  }

  String createJwt(final String user, final List<String> roles) {
    return jwtAuthHelper.createJwt(
        user,
        List.of("read", "write"),
        roles,
        Duration.ofHours(1),
        UUID.randomUUID().toString()
    );
  }
}

File: src/test/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/controllers/ResourceTest.kt
package uk.gov.justice.digital.hmpps.riskprofiler.controllers

import org.assertj.core.api.Assertions
import org.awaitility.kotlin.await
import org.awaitility.kotlin.until
import org.junit.jupiter.api.BeforeEach
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.boot.test.json.JsonContent
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.core.ResolvableType
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.reactive.server.WebTestClient
import uk.gov.justice.digital.hmpps.riskprofiler.dao.DataRepositoryFactory
import uk.gov.justice.digital.hmpps.riskprofiler.integration.wiremock.OAuthMockServer.Companion.oauthMockServer
import uk.gov.justice.digital.hmpps.riskprofiler.integration.mocks.PathfinderMockServer.Companion.pathfinderMockServer
import uk.gov.justice.digital.hmpps.riskprofiler.integration.mocks.PrisonMockServer.Companion.prisonMockServer
import uk.gov.justice.digital.hmpps.riskprofiler.services.NomisService
import uk.gov.justice.digital.hmpps.riskprofiler.utils.JwtAuthenticationHelper
import java.time.Duration
import java.util.Objects
import java.util.UUID

@ActiveProfiles(profiles = ["test", "localstack"])
@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
abstract class ResourceTest {

  @Autowired
  protected lateinit var dataRepositoryFactory: DataRepositoryFactory

  @Autowired
  protected lateinit var testRestTemplate: TestRestTemplate

  @Autowired
  protected lateinit var webTestClient: WebTestClient

  @Autowired
  protected lateinit var jwtAuthHelper: JwtAuthenticationHelper

  @Autowired
  protected lateinit var nomisService: NomisService

  fun createHttpEntityWithBearerAuthorisation(user: String?, roles: List<String>?): HttpEntity<*> {
    val jwt = createJwt(user, roles)
    return createHttpEntity(jwt, null)
  }

  private fun allFilesLoaded(): Boolean {
    return dataRepositoryFactory.getRepositories().stream().allMatch { it.dataAvailable() }
  }

  @BeforeEach
  fun resetStubs() {
    nomisService.evictSocListAlertsCache("A1234AB")
    nomisService.evictSocListAlertsCache("A1234AC")
    nomisService.evictSocListAlertsCache("A1234AE")
    nomisService.evictSocListAlertsCache("A5015DY")
    nomisService.evictIncidentsCache("A1234AB")
    nomisService.evictIncidentsCache("A1234AC")
    nomisService.evictEscapeListAlertsCache("A1234AB")
    nomisService.evictEscapeListAlertsCache("A1234AC")

    prisonMockServer.resetAll()
    oauthMockServer.resetAll()
    pathfinderMockServer.resetAll()
    await until { allFilesLoaded() }
  }

  fun createHttpEntity(bearerToken: String, body: Any?): HttpEntity<*> {
    val headers = HttpHeaders()
    headers.add(HttpHeaders.AUTHORIZATION, "Bearer $bearerToken")
    headers.add(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE)
    if (body != null) {
      headers.add(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
    }
    return HttpEntity(body, headers)
  }

  fun assertThatStatus(response: ResponseEntity<String?>, status: Int) {
    Assertions.assertThat(response.statusCode).withFailMessage(
      "Expecting status code value <%s> to be equal to <%s> but it was not.\nBody was\n%s",
      response.statusCode,
      status,
      response.body
    ).isEqualTo(status)
  }

  fun assertThatJsonFileAndStatus(response: ResponseEntity<String?>, status: Int, jsonFile: String?) {
    assertThatStatus(response, status)
    Assertions.assertThat(getBodyAsJsonContent<Any>(response)).isEqualToJson(jsonFile)
  }

  private fun <T> getBodyAsJsonContent(response: ResponseEntity<String?>): JsonContent<T> {
    return JsonContent(javaClass, ResolvableType.forType(String::class.java), Objects.requireNonNull(response.body))
  }

  fun createJwt(user: String?, roles: List<String>?): String {
    return jwtAuthHelper.createJwt(
      user,
      java.util.List.of("read", "write"),
      roles,
      Duration.ofHours(1),
      UUID.randomUUID().toString()
    )
  }
}
